tls-cipher is not parsed

openvpn_server_config_parser.exe windows-riseup-client.cfg 
openvpn_server_config_parser.exe: [INFO] Reading file "windows-riseup-client.cfg"
openvpn_server_config_parser.exe: [ERROR] : Error at byte offset 988: "tls-cipher TLS-DHE-RSA-WITH-AES-256-CBC-SHA\ncipher AES-256-CBC\nauth SHA256\n\ndev tun\n# if dev tun doe"

--- a/src/config.ml
+++ b/src/config.ml
@@ -39,6 +39,52 @@ let int_of_ping_interval = function
          violated"
   | `Seconds n -> n
 
+let (cipher_to_cs : string -> Tls.Ciphersuite.ciphersuite),
+    (cs_to_cipher : Tls.Ciphersuite.ciphersuite -> string) =
+  let map = [
+    "TLS-ECDHE-ECDSA-WITH-AES-256-GCM-SHA384", `ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 ;
+    "TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384", `ECDHE_RSA_WITH_AES_256_GCM_SHA384 ;
+    "TLS-DHE-RSA-WITH-AES-256-GCM-SHA384", `DHE_RSA_WITH_AES_256_GCM_SHA384 ;
+    "TLS-ECDHE-ECDSA-WITH-CHACHA20-POLY1305-SHA256", `ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256 ;
+    "TLS-ECDHE-RSA-WITH-CHACHA20-POLY1305-SHA256", `ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 ;
+    "TLS-DHE-RSA-WITH-CHACHA20-POLY1305-SHA256", `DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 ;
+    "TLS-ECDHE-ECDSA-WITH-AES-128-GCM-SHA256", `ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 ;
+    "TLS-ECDHE-RSA-WITH-AES-128-GCM-SHA256", `ECDHE_RSA_WITH_AES_128_GCM_SHA256 ;
+    "TLS-DHE-RSA-WITH-AES-128-GCM-SHA256", `DHE_RSA_WITH_AES_128_GCM_SHA256 ;
+    "TLS-ECDHE-ECDSA-WITH-AES-256-CBC-SHA384", `ECDHE_ECDSA_WITH_AES_256_CBC_SHA384 ;
+    "TLS-ECDHE-RSA-WITH-AES-256-CBC-SHA384", `ECDHE_RSA_WITH_AES_256_CBC_SHA384 ;
+    "TLS-DHE-RSA-WITH-AES-256-CBC-SHA256", `DHE_RSA_WITH_AES_256_CBC_SHA256 ;
+    "TLS-ECDHE-ECDSA-WITH-AES-128-CBC-SHA256", `ECDHE_ECDSA_WITH_AES_128_CBC_SHA256 ;
+    "TLS-ECDHE-RSA-WITH-AES-128-CBC-SHA256", `ECDHE_RSA_WITH_AES_128_CBC_SHA256 ;
+    "TLS-DHE-RSA-WITH-AES-128-CBC-SHA256", `DHE_RSA_WITH_AES_128_CBC_SHA256 ;
+    "TLS-ECDHE-ECDSA-WITH-AES-256-CBC-SHA", `ECDHE_ECDSA_WITH_AES_256_CBC_SHA ;
+    "TLS-ECDHE-RSA-WITH-AES-256-CBC-SHA", `ECDHE_RSA_WITH_AES_256_CBC_SHA ;
+    "TLS-DHE-RSA-WITH-AES-256-CBC-SHA", `DHE_RSA_WITH_AES_256_CBC_SHA ;
+    "TLS-ECDHE-ECDSA-WITH-AES-128-CBC-SHA", `ECDHE_ECDSA_WITH_AES_128_CBC_SHA ;
+    "TLS-ECDHE-RSA-WITH-AES-128-CBC-SHA", `ECDHE_RSA_WITH_AES_128_CBC_SHA ;
+    "TLS-DHE-RSA-WITH-AES-128-CBC-SHA", `DHE_RSA_WITH_AES_128_CBC_SHA ;
+  ] in
+  let rev_map = List.map (fun (a, b) -> b, a) map in
+  (fun s ->
+     match List.assoc_opt s map with
+     | None -> failwith ("unsupported tls-cipher " ^ s)
+     | Some c -> c),
+  (fun c -> List.assoc c rev_map)
+
+let (cipher13_to_cs13 : string -> Tls.Ciphersuite.ciphersuite13),
+    (cs13_to_cipher13 : Tls.Ciphersuite.ciphersuite13 -> string) =
+  let map = [
+    "TLS_AES_256_GCM_SHA384", `AES_256_GCM_SHA384 ;
+    "TLS_CHACHA20_POLY1305_SHA256", `CHACHA20_POLY1305_SHA256 ;
+    "TLS_AES_128_GCM_SHA256", `AES_128_GCM_SHA256 ;
+  ] in
+  let rev_map = List.map (fun (a, b) -> b, a) map in
+  (fun s ->
+     match List.assoc_opt s map with
+     | None -> failwith ("unsupported tls-ciphersuite " ^ s)
+     | Some c -> c),
+  (fun c -> List.assoc c rev_map)
+
 module Conf_map = struct
   type flag = unit
 
@@ -138,7 +184,9 @@ module Conf_map = struct
     | Tls_mode : [ `Client | `Server ] k
     | Tls_key : X509.Private_key.t k
     | Tls_timeout : int k
-    | Tls_version_min : ([ `V1_3 | `V1_2 | `V1_1 ] * bool) k
+    | Tls_version_min : (Tls.Core.tls_version * bool) k
+    | Tls_cipher : Tls.Ciphersuite.ciphersuite list k
+    | Tls_ciphersuite : Tls.Ciphersuite.ciphersuite13 list k
     | Topology : [ `Net30 | `P2p | `Subnet ] k
     | Transition_window : int k
     | Tun_mtu : int k
@@ -396,11 +444,17 @@ module Conf_map = struct
     | Tls_key, key -> pp_x509_private_key key
     | Tls_version_min, (ver, or_highest) ->
         p () "tls-version-min %s%s"
-          (match ver with `V1_3 -> "1.3" | `V1_2 -> "1.2" | `V1_1 -> "1.1")
+          (match ver with `TLS_1_3 -> "1.3" | `TLS_1_2 -> "1.2" | `TLS_1_1 -> "1.1" | `TLS_1_0 -> "1.0")
           (if or_highest then " or-highest" else "")
     | Tls_mode, `Server -> p () "tls-server"
     | Tls_mode, `Client -> p () "tls-client"
     | Tls_timeout, seconds -> p () "tls-timeout %d" seconds
+    | Tls_cipher, ciphers ->
+      p () "tls-cipher %a" Fmt.(list ~sep:(any ":") string)
+        (List.map cs_to_cipher ciphers)
+    | Tls_ciphersuite, ciphers ->
+      p () "tls-ciphersuite %a" Fmt.(list ~sep:(any ":") string)
+        (List.map cs13_to_cipher13 ciphers)
     | Topology, v ->
         p () "topology %s"
           (match v with
@@ -709,6 +763,16 @@ let a_tls_auth_payload direction =
   inline_payload "TLS AUTH" >>| fun (a, b, c, d) ->
   B (Tls_auth, (direction, a, b, c, d))
 
+let split_colon s = String.split_on_char ':' s
+
+let a_tls_cipher =
+  string "tls-cipher" *> a_whitespace *> a_line not_control_char >>| fun ciphers ->
+  `Entry (B (Tls_cipher, List.map cipher_to_cs (split_colon ciphers)))
+
+let a_tls_ciphersuite =
+  string "tls-ciphersuite" *> a_whitespace *> a_line not_control_char >>| fun ciphers ->
+  `Entry (B (Tls_ciphersuite, List.map cipher13_to_cs13 (split_colon ciphers)))
+
 let a_auth_user_pass =
   a_option_with_single_path "auth-user-pass" `Auth_user_pass
 
@@ -845,9 +909,10 @@ let a_tls_version_min =
   string "tls-version-min" *> a_whitespace
   *> choice
        [
-         string "1.3" *> return `V1_3;
-         string "1.2" *> return `V1_2;
-         string "1.1" *> return `V1_1;
+         string "1.3" *> return `TLS_1_3;
+         string "1.2" *> return `TLS_1_2;
+         string "1.1" *> return `TLS_1_1;
+         string "1.0" *> return `TLS_1_0;
        ]
   >>= fun v ->
   choice
@@ -1262,6 +1327,8 @@ let a_config_entry : line A.t =
          a_route;
          a_route_gateway;
          a_topology;
+         a_tls_ciphersuite;
+         a_tls_cipher;
          a_not_implemented;
          a_whitespace *> return (`Ignored "");
        ]

--- a/src/engine.ml
+++ b/src/engine.ml
@@ -433,8 +433,23 @@ let incoming_control_client config rng session channel now op data =
           match (Config.find Tls_cert config, Config.find Tls_key config) with
           | Some cert, Some key -> `Single ([ cert ], key)
           | _ -> `None
+        and ciphers =
+          match Config.find Tls_cipher config, Config.find Tls_ciphersuite config with
+          | Some c, None -> Some c
+          | None, Some c -> Some (c :> Tls.Ciphersuite.ciphersuite list)
+          | Some c, Some c' -> Some (c @ (c' :> Tls.Ciphersuite.ciphersuite list))
+          | None, None -> None
+        and version =
+          match Config.find Tls_version_min config with
+          | None -> None
+          | Some (v, or_highest) ->
+            let highest = `TLS_1_3 in (* update when ocaml-tls supports new versions *)
+            if or_highest then
+              Some (highest, highest)
+            else
+              Some (v, highest)
         in
-        Tls.(Engine.client (Config.client ~certificates ~authenticator ()))
+        Tls.(Engine.client (Config.client ?ciphers ?version ~certificates ~authenticator ()))
       in
       Ok
         ( None,
@@ -522,8 +537,24 @@ let incoming_control_server is_not_taken config rng session channel _now _ts
           Config.get Tls_cert config,
           Config.get Tls_key config )
       in
+      let ciphers =
+        match Config.find Tls_cipher config, Config.find Tls_ciphersuite config with
+          | Some c, None -> Some c
+          | None, Some c -> Some (c :> Tls.Ciphersuite.ciphersuite list)
+          | Some c, Some c' -> Some (c @ (c' :> Tls.Ciphersuite.ciphersuite list))
+          | None, None -> None
+      and version =
+        match Config.find Tls_version_min config with
+        | None -> None
+        | Some (v, or_highest) ->
+          let highest = `TLS_1_3 in (* update when ocaml-tls supports new versions *)
+          if or_highest then
+            Some (highest, highest)
+          else
+            Some (v, highest)
+      in
       let tls_config =
-        Tls.Config.server ~certificates:(`Single ([ server ], key)) ()
+        Tls.Config.server ?ciphers ?version ~certificates:(`Single ([ server ], key)) ()
       in
       let tls = Tls.Engine.server tls_config in
       let channel = { channel with channel_st = TLS_handshake tls } in

--- a/src/miragevpn.mli
+++ b/src/miragevpn.mli
@@ -181,9 +181,13 @@ module Config : sig
     (** Retransmit control channel packet after not receiving an ACK for
         [n] seconds. TODO: presumably does not apply to TCP? *)
 
-    | Tls_version_min : ([`V1_3 | `V1_2 | `V1_1 ] * bool) k
+    | Tls_version_min : (Tls.Core.tls_version * bool) k
     (** [v * or_highest]: if [or_highest] then v = the highest version supported
         by the TLS library. *)
+
+    | Tls_cipher : Tls.Ciphersuite.ciphersuite list k
+
+    | Tls_ciphersuite : Tls.Ciphersuite.ciphersuite13 list k
 
     | Topology : [`Net30 | `P2p | `Subnet] k
 

--- a/test/config_tests.ml
+++ b/test/config_tests.ml
@@ -488,6 +488,12 @@ let tls_home_conf =
   |> add_b (a_key_payload (string_of_file "client.key"))
   |> add Cipher "AES-256-CBC" |> add Verb 3
 
+let tls_home_conf_with_cipher =
+  let open Miragevpn.Config in
+  tls_home_conf
+  |> add Tls_cipher [ `ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 ]
+  |> add Tls_ciphersuite [ `CHACHA20_POLY1305_SHA256 ]
+
 let ipredator_conf =
   let ca =
     match
@@ -575,7 +581,7 @@ efabaa5e34619f13adbe58b6c83536d3
   |> add Replay_window (512, 60)
   |> add Mute_replay_warnings ()
   |> add Ifconfig_nowarn ()
-  |> add Tls_version_min (`V1_2, false)
+  |> add Tls_version_min (`TLS_1_2, false)
 
 let parse_multiple_cas () =
   let file = "multi-ca-client.conf" in
@@ -633,7 +639,10 @@ let tests =
       parse_multiple_cas );
     (* "parsing configuration 'wild-client'", `Quick,
        parse_client_configuration "wild-client.conf" ; -- verify-x509-name *)
-    (* "parsing configuration 'windows-riseup-client'", `Quick,
-       parse_client_configuration "windows-riseup-client.conf" ; -- tls-cipher *)
+    (* ( "parsing configuration 'windows-riseup-client'", `Quick,
+       parse_client_configuration "windows-riseup-client.conf" ); --auth *)
+    ( "parsing 'tls-home-with-cipher'",
+      `Quick,
+      parse_client_configuration ~config:tls_home_conf_with_cipher "tls-home-with-cipher.conf" );
     ("crowbar fuzzing", `Slow, crowbar_fuzz_config);
   ]

--- a/test/sample-configuration-files/tls-home-with-cipher.conf
+++ b/test/sample-configuration-files/tls-home-with-cipher.conf
@@ -1,0 +1,88 @@
+#
+# Sample OpenVPN configuration file for
+# home using SSL/TLS mode and RSA certificates/keys.
+#
+# '#' or ';' may be used to delimit comments.
+
+# Use a dynamic tun device.
+# For Linux 2.2 or non-Linux OSes,
+# you may want to use an explicit
+# unit number such as "tun1".
+# OpenVPN also supports virtual
+# ethernet "tap" devices.
+dev tun
+
+# Our OpenVPN peer is the office gateway.
+remote 1.2.3.4
+
+# 10.1.0.2 is our local VPN endpoint (home).
+# 10.1.0.1 is our remote VPN endpoint (office).
+ifconfig 10.1.0.2 10.1.0.1
+
+# Our up script will establish routes
+# once the VPN is alive.
+up ./home.up
+
+# In SSL/TLS key exchange, Office will
+# assume server role and Home
+# will assume client role.
+tls-client
+
+# Certificate Authority file
+ca ca.crt
+
+# Our certificate/public key
+cert client.crt
+
+# Our private key
+key client.key
+
+# OpenVPN 2.0 uses UDP port 1194 by default
+# (official port assignment by iana.org 11/04).
+# OpenVPN 1.x uses UDP port 5000 by default.
+# Each OpenVPN tunnel must use
+# a different port number.
+# lport or rport can be used
+# to denote different ports
+# for local and remote.
+; port 1194
+
+# Downgrade UID and GID to
+# "nobody" after initialization
+# for extra security.
+; user nobody
+; group nobody
+
+# If you built OpenVPN with
+# LZO compression, uncomment
+# out the following line.
+; comp-lzo
+
+# Send a UDP ping to remote once
+# every 15 seconds to keep
+# stateful firewall connection
+# alive.  Uncomment this
+# out if you are using a stateful
+# firewall.
+; ping 15
+
+# Uncomment this section for a more reliable detection when a system
+# loses its connection.  For example, dial-ups or laptops that
+# travel to other locations.
+; ping 15
+; ping-restart 45
+; ping-timer-rem
+; persist-tun
+; persist-key
+
+cipher AES-256-CBC
+
+# Verbosity level.
+# 0 -- quiet except for fatal errors.
+# 1 -- mostly quiet, but display non-fatal network errors.
+# 3 -- medium output, good for normal operation.
+# 9 -- verbose, good for troubleshooting
+verb 3
+
+tls-cipher TLS-ECDHE-ECDSA-WITH-AES-256-GCM-SHA384
+tls-ciphersuite TLS_CHACHA20_POLY1305_SHA256

